plugins/realtek-mst/fu-realtek-mst-device.c

```diff
@@ -0,0 +1,910 @@
+	if (!mst_read_register_indirect (self, REG_GPIO88_CONFIG, &value, error))
+		return FALSE;
+	if (!mst_write_register_indirect (self,
+					  REG_GPIO88_CONFIG,
+					  (value & 0xF0) | 1,
+					  error))
+		return FALSE;
+
+	/* set output level */
+	g_debug ("set pin 88 = %d", level);
+	if (!mst_read_register_indirect (self, REG_GPIO88_VALUE, &value, error))
+		return FALSE;
+	return mst_write_register_indirect (self, REG_GPIO88_VALUE,
+					    (value & 0xFE) | (level != FALSE),
+					    error);
+}
+
+static gboolean
+fu_realtek_mst_device_probe (FuDevice *device, GError **error)
+{
+	FuRealtekMstDevice *self = FU_REALTEK_MST_DEVICE (device);
+	FuContext *context = fu_device_get_context (device);
+	const gchar *quirk_name = NULL;
+	const gchar *hardware_family = NULL;
+	g_autofree gchar *physical_id = NULL;
+	g_autofree gchar *instance_id = NULL;
+	g_autofree gchar *family_instance_id = NULL;
+
+	if (!FU_DEVICE_CLASS (fu_realtek_mst_device_parent_class)->probe (device, error))
+		return FALSE;
+
+	physical_id = g_strdup_printf ("I2C_PATH=%s",
+				       fu_udev_device_get_sysfs_path (
+					       FU_UDEV_DEVICE (device)));
+	fu_device_set_physical_id (device, physical_id);
+
+	/* set custom instance ID and load matching quirks */
+	instance_id = g_strdup_printf ("REALTEK-MST\\Name_%s",
+				       fu_udev_device_get_sysfs_attr (
+					       FU_UDEV_DEVICE (device),
+					       "name",
+					       NULL));
+	fu_device_add_instance_id (device, instance_id);
+
+	hardware_family = fu_context_get_hwid_value (context, FU_HWIDS_KEY_FAMILY);
+	family_instance_id = g_strdup_printf ("%s&Family_%s", instance_id, hardware_family);
+	fu_device_add_instance_id_full (device, family_instance_id,
+					FU_DEVICE_INSTANCE_FLAG_ONLY_QUIRKS);
+
+	/* having loaded quirks, check this device is supported */
+	quirk_name = fu_device_get_name (device);
+	if (g_strcmp0 (quirk_name, "RTD2142") != 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "only RTD2142 is supported");
+		return FALSE;
+	}
+
+	if (self->dp_aux_dev_name == NULL) {
+		g_set_error_literal (error, FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "RealtekMstDpAuxName must be specified");
+		return FALSE;
+	}
+
+	if ((self->bus_device = fu_realtek_mst_device_locate_bus (self, error)) == NULL)
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_realtek_mst_device_open (FuDevice *device, GError **error)
+{
+	FuRealtekMstDevice *self = FU_REALTEK_MST_DEVICE (device);
+	const gchar *bus_path = fu_udev_device_get_device_file (self->bus_device);
+	gint bus_fd;
+
+	/* open the bus and not self */
+	if ((bus_fd = g_open (bus_path, O_RDWR)) == -1) {
+		g_set_error (error, G_IO_ERROR,
+#ifdef HAVE_ERRNO_H
+			     g_io_error_from_errno (errno),
+#else
+	       		     G_IO_ERROR_FAILED,
+#endif
+			     "failed to open %s", bus_path);
+
+		return FALSE;
+	}
+	fu_udev_device_set_fd (FU_UDEV_DEVICE (self), bus_fd);
```

Comment:
I think this probably leaks FDs like a sieve. I'll port the module to FuI2cDevice which should fix things up.

Reply:
`fu_udev_device_set_fd` closes anything that's already open, so I think it's fine.

Reply:
Ohh it's more that you're opening the bus device and don't seem to be closing it anywhere. I think we can just remove the bus device altogether. one sec.

Reply:
I've pushed a few things to fix up nits and port to FuI2cDevice, sorry if I broke everything.